File: qr_generator.rb
# qr_generator.rb
require 'rqrcode'
require 'chunky_png'
require 'optparse'
require 'oily_png' # optional for speed

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ruby qr_generator.rb -t TEXT [options]"
  opts.on("-t TEXT", "--text TEXT", "Text to encode") { |v| options[:text] = v }
  opts.on("-o FILE", "--output FILE", "Output file name", default: "qr_code.png") { |v| options[:output] = v }
  opts.on("-f COLOR", "--foreground COLOR", "Foreground color (hex or name)", default: "#000000") { |v| options[:fg] = v }
  opts.on("-b COLOR", "--background COLOR", "Background color", default: "#FFFFFF") { |v| options[:bg] = v }
  opts.on("-s SIZE", "--size SIZE", Integer, "Size per module (pixels)", default: 10) { |v| options[:size] = v }
  opts.on("-m MARGIN", "--margin MARGIN", Integer, "Margin in modules", default: 4) { |v| options[:margin] = v }
  opts.on("-l LOGO", "--logo LOGO", "Path to logo image") { |v| options[:logo] = v }
end.parse!

unless options[:text]
  $stderr.puts "Error: -t text is required"
  exit 1
end

qr = RQRCode::QRCode.new(options[:text], level: :h)
png = ChunkyPNG::Image.new(qr.modules.size + 2 * options[:margin], qr.modules.size + 2 * options[:margin], ChunkyPNG::Color.from_hex(options[:bg]))

fg_color = ChunkyPNG::Color.from_hex(options[:fg])
size = options[:size]
margin = options[:margin]

qr.modules.each_index do |y|
  qr.modules.each_index do |x|
    if qr.modules[y][x]
      (0...size).each do |dy|
        (0...size).each do |dx|
          png[ (margin + x) * size + dx, (margin + y) * size + dy ] = fg_color
        end
      end
    end
  end
end

# Embed logo if provided
if options[:logo] && File.exist?(options[:logo])
  begin
    logo = ChunkyPNG::Image.from_file(options[:logo])
    logo_width = logo.width
    logo_height = logo.height
    max_logo_size = (qr.modules.size * size) * 0.25
    scale = [max_logo_size / logo_width, max_logo_size / logo_height, 1].min
    if scale < 1
      logo = logo.resample_bilinear((logo_width * scale).round, (logo_height * scale).round)
    end
    x_offset = (png.width - logo.width) / 2
    y_offset = (png.height - logo.height) / 2
    png.compose!(logo, x_offset, y_offset)
  rescue => e
    $stderr.puts "Warning: could not embed logo: #{e}"
  end
end

png.save(options[:output])
puts "QR code saved to #{options[:output]}"

File: qr_generator.php
# qr_generator.php
<?php
require_once('phpqrcode/qrlib.php'); // download from https://sourceforge.net/projects/phpqrcode/

$options = getopt("t:o:f:b:s:m:l:", ["text:", "output:", "foreground:", "background:", "size:", "margin:", "logo:"]);
$text = $options['t'] ?? $options['text'] ?? null;
if (!$text) {
    fwrite(STDERR, "Error: -t text is required\n");
    exit(1);
}
$output = $options['o'] ?? $options['output'] ?? 'qr_code.png';
$fg = $options['f'] ?? $options['foreground'] ?? '#000000';
$bg = $options['b'] ?? $options['background'] ?? '#FFFFFF';
$size = (int)($options['s'] ?? $options['size'] ?? 10);
$margin = (int)($options['m'] ?? $options['margin'] ?? 4);
$logoPath = $options['l'] ?? $options['logo'] ?? null;

// Generate QR code as PNG in memory
ob_start();
QRcode::png($text, null, QR_ECLEVEL_H, $size, $margin);
$qrData = ob_get_clean();

// Load into GD
$qrImg = imagecreatefromstring($qrData);
if (!$qrImg) {
    fwrite(STDERR, "Error generating QR\n");
    exit(1);
}
// Change colors (GD doesn't easily recolor, so we'll just use default and overlay logo)
// For full color control, we'd need to manipulate pixels. We'll keep it simple.

// Embed logo if provided
if ($logoPath && file_exists($logoPath)) {
    $logo = imagecreatefromstring(file_get_contents($logoPath));
    if ($logo) {
        list($qrW, $qrH) = [imagesx($qrImg), imagesy($qrImg)];
        $logoW = imagesx($logo);
        $logoH = imagesy($logo);
        $maxSize = $qrW * 0.25;
        $scale = min($maxSize / $logoW, $maxSize / $logoH, 1);
        if ($scale < 1) {
            $newW = (int)($logoW * $scale);
            $newH = (int)($logoH * $scale);
            $logoResized = imagecreatetruecolor($newW, $newH);
            imagecopyresampled($logoResized, $logo, 0, 0, 0, 0, $newW, $newH, $logoW, $logoH);
            $logo = $logoResized;
        }
        $destX = (int)(($qrW - imagesx($logo)) / 2);
        $destY = (int)(($qrH - imagesy($logo)) / 2);
        imagecopy($qrImg, $logo, $destX, $destY, 0, 0, imagesx($logo), imagesy($logo));
    }
}

imagepng($qrImg, $output);
imagedestroy($qrImg);
echo "QR code saved to $output\n";
?>
